R/parallel.R

```diff
@@ -1,41 +1,102 @@
 
-#' Create cluster
+#' Create a rbmi ready cluster
 #'
-#' @param ncores Number of parallel processes to use
+#' @param ncores Number of parallel processes to use or an existing cluster to make use of
+#' @param objects a named list of objects to export into the sub-processes
+#' @param packages a character vector of libraries to load in the sub-processes
 #'
-#' If `ncores` is `1` this function will return NULL
-#' This function spawns a PSOCK cluster.
-#' Ensures that `rbmi` and `assert_that` have been loaded
-#' on the sub-processes
+#' This function is a wrapper around `parallel::makePSOCKcluster()` but takes
+#' care of configuring rbmi to be used in the sub-processes as well as loading
+#' user defined objects and libraries and setting the seed for reproducibility.
 #'
-get_cluster <- function(ncores = 1) {
-    if (ncores == 1) {
+#' If `ncores` is `1` this function will return `NULL`.
+#'
+#' If `ncores` is a cluster created via `parallel::makeCluster()` then this function
+#' just takes care of inserting the relevant rbmi objects into the existing cluster.
+#'
+#' @examples
+#' \dontrun{
+#' # Basic usage
+#' make_rbmi_cluster(5)
+#'
+#' # User objects + libraries
+#' VALUE <- 5
+#' myfun <- function(x) {
+#'     x + day(VALUE) # From lubridate::day()
+#' }
+#' make_rbmi_cluster(5, list(VALUE = VALUE, myfun = myfun), c("lubridate"))
+#'
+#' # Using a already created cluster
+#' cl <- parallel::makeCluster(5)
+#' make_rbmi_cluster(cl)
+#' }
+#' @export
+make_rbmi_cluster <- function(ncores = 1, objects = NULL, packages = NULL) {
+
+    if (is.numeric(ncores) && ncores == 1) {
         return(NULL)
+    } else if (is.numeric(ncores)) {
+        cl <- parallel::makePSOCKcluster(ncores)
+    } else if (is(ncores, "cluster")) {
+        cl <- ncores
+    } else {
+        stop(sprintf(
+            "`ncores` has unsupported class of: %s",
+            paste(class(ncores), collapse = ", ")
+        ))
+    }
+
+    # Load user defined objects into the globalname space
+    if (!is.null(objects) && length(objects)) {
+        export_env <- list2env(objects)
+        parallel::clusterExport(cl, names(objects), export_env)
     }
 
-    cl <- parallel::makePSOCKcluster(
-        ncores
+    # Load user defined packages
+    packages <- if (is.null(packages)) {
+        # TODO - can't remember why this is needed; need to look into
+        "assertthat"
+    } else {
+        c(packages, "assertthat")
+    }
```

Comment:
This would be the same I think if packages is NULL or a character vector
```suggestion
    packages <- c(packages, "assertthat")
```